Add multi-pass environment variable resolution

john-patterson opened the pull request:
Per issue [#2057](https://github.com/Microsoft/vscode-cpptools/issues/2057).

This allows a user to nest environment tags, or define tags based on the definition of other environment variables. If the result of a resolution still contains a `${...}` tag, we will continue to scan the input until we reach a steady state where the variable expansion results in the same input.

This should allow the following forms which were previously not possible in the environment definition:

* Dependent variable definition
```
"env": {
    "envRoot": "apps/tool/builldenv",
    "arm6.include": "${envRoot}/arm6/include"
},
...
"some_config": "${arm6.include}
```

* Nested variable definition of any degree
```
"env": {
    "root": "apps/tool",
    "armVersion": "${config:armVersion}",
    "arm5.include": "${root}/legacy/arm/include",
    "arm6.include": "${root}/arm/include",
    "armTestPath": "${arm${armVersion}.include}/test"
}
```

* Additionally, this fixes an open bug that was not reported. If a variable contained "env", "config", or "workspaceFolder" it would not be parsed correctly. If you used a variable ${envRoot} it would match "envR" as the type and "oot" as the variable. This has been resolved.

Commit:
fix unit test launch.json target

We were passing the root of the test folder instead of the unitTests
specific folder into the runner, so it wasn't finding unit tests.

Commit:
fix res. of vars with env,config,workspaceFolder

If a variable had env, config, or workspaceFolder as substring, it was
being treated as an environment variable lookup on the reminder of the
string minus the character immediately following the substring. For
instance: envRoot was being looked up as "oot" in the configuarion
environment block.

The issue is that the regex to match the env., config., and
workspaceFolder. patterns used . instead of \.

john-patterson committed:
tests existing tag resolution functionality

john-patterson committed:
rough draft

john-patterson committed:
refactors common test file

This gets rid of linting errors and adopts a more BDD literate assert style.

john-patterson committed:
fixes linting errors in common.ts

Commit:
refactors resolveVariables implementation

Commit:
adds specific test for user scenario

Commit:
fix cycle detection and half open detection

sean-mcmanus commented on Extension/.vscode/launch.json:
```diff
@@ -24,7 +24,7 @@
 			"args": [
 				"--extensionDevelopmentPath=${workspaceFolder}",
-				"--extensionTestsPath=${workspaceFolder}/out/test"
+				"--extensionTestsPath=${workspaceFolder}/out/test/unitTests"
```
Oh, that's interesting...I guess we don't Launch Tests very much :)

sean-mcmanus replied:
Looks good so far. I'm not done reviewing yet...

john-patterson committed:
simplifies parsing into single pass

sean-mcmanus commented on Extension/src/common.ts:
```diff
@@ -178,54 +178,59 @@ export function resolveVariables(input: string, additionalEnvironment: {[key: st
-                    newValue = process.env[name];
-                }
-                break;
+    let regexp: () => RegExp = () => /\$\{((env|config|workspaceFolder)(\.|:))?(.*?)\}/g;
```
Good catch 👍

sean-mcmanus reviewed (approved):
Seems good to me. Bob will probably want to finish reviewing as well.

john-patterson committed:
fixes Travis CI failing unitTests

As per https://github.com/Microsoft/vscode-wordcount/issues/5, running tests from CLI isn't supported with vscode except through the [workaround listed for Travis CI](https://code.visualstudio.com/docs/extensions/testing-extensions#_running-tests-automatically-on-travis-ci-build-machines).
